lib/tree.rb

```diff
@@ -16,51 +18,264 @@ def initialize
   def find(key)
-    raise NotImplementedError
+    return nil if @root.nil?
+
+    current = @root
+    until current.nil?
+      return current.value if current.key == key
+
+      current = if key > current.key
+                  current.right
+                else
+                  current.left
+                end
+    end
+
+    return nil
+  end
+
+  def find_return_node(key)
+    return nil if @root.nil?
+
+    current = @root
+    until current.nil?
+      return current if current.key == key
+
+      current = if key > current.key
+                  current.right
+                else
+                  current.left
+                end
+    end
+
+    return nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # >> hit every node in tree during traversal
+  # Space Complexity: O(log n)
+  # >> go down one side of tree at a time --> stack has height # of levels
+  # >> O(n) if unbalanced
   def inorder
-    raise NotImplementedError
+    return [] if @root.nil?
+
+    nodes_arr = []
+    inorder_helper(@root, nodes_arr)
+
+    return nodes_arr
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def inorder_helper(current, nodes_arr)
+    return if current.nil?
+
+    inorder_helper(current.left, nodes_arr)
+    nodes_arr << { key: current.key, value: current.value }
+    inorder_helper(current.right, nodes_arr)
+  end
+
+  # Time Complexity: O(n)
+  # >> hit every node in tree during traversal
+  # Space Complexity: O(log n)
+  # >> go down one side of tree at a time --> stack has height # of levels
+  # >> O(n) if unbalanced
   def preorder
-    raise NotImplementedError
+    return [] if @root.nil?
+
+    nodes_arr = []
+    preorder_helper(@root, nodes_arr)
+
+    return nodes_arr
+  end
+
+  def preorder_helper(current, nodes_arr)
+    return if current.nil?
+
+    nodes_arr << { key: current.key, value: current.value }
+    preorder_helper(current.left, nodes_arr)
+    preorder_helper(current.right, nodes_arr)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # >> hit every node in tree during traversal
+  # Space Complexity: O(log n)
+  # >> go down one side of tree at a time --> stack has height # of levels
+  # >> O(n) if unbalanced
   def postorder
-    raise NotImplementedError
+    return [] if @root.nil?
+
+    nodes_arr = []
+    postorder_helper(@root, nodes_arr)
+
+    return nodes_arr
+  end
+
+  def postorder_helper(current, nodes_arr)
+    return if current.nil?
+
+    postorder_helper(current.left, nodes_arr)
+    postorder_helper(current.right, nodes_arr)
+    nodes_arr << { key: current.key, value: current.value }
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # >> Go through every node in tree to confirm deepest level
+  # Space Complexity: O(log n)
+  # >> store set of unique levels / depths (log n)
+  # >> store current node O(1)
+  # >> store current height O(1)
   def height
```

Comment:
👍 Correct on time complexity and the space complexity is correct if the tree is balanced.